frontend/alert/src/Timeline.js

```diff
@@ -0,0 +1,222 @@
+  height: 20px;
+  border-radius: 3px;
+  font-weight: 600;
+  color: #e8746a;
+  background: #f9dcda;
+`
+
+const OpenBadge = styled(ClosedBadge)`
+  color:
+`
+
+const Segment = styled.div`
+  width: 2px;
+  height: ${props=>props.height + 5}px;
+  background: ${props=>props.type=="opened" ? "#78d381" : "#cbd5dd"};
+`
+const MyCircle = styled.div`
+  color: ${props=>props.type=="opened" ? "#78d381" : "#cbd5dd"};
+`
+
+const TimeStyle = styled.div`
+  position: absolute;
+  top: ${props => props.offset}px;
+  right: 50px;
+`
+
+const DateStyle = styled.div`
+  position: absolute;
+  top: ${props => props.offset}px;
+  left: 10px;
+`
+
+const CourseIndicator = ({time, type, offset}) => {
+  let convertedTime = convertTime(time);
+  return <TimeStyle offset={offset}>{convertedTime[1]}</TimeStyle>
+}
+
+const FlexRow = styled.div`
+  display: flex;
+  flex-direction: row;
+  position: relative;
+`
+
+const LeftRight = styled.div`
+  display: flex;
+  flex-direction: row;
+  justify-content: space-between;
+  align-items: center;
+`
+
+function convertTime(timeString){
+  const months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
+  let d = new Date(timeString);
+  return [d.toLocaleDateString('en-US', {month:"numeric", day: "numeric"}), d.toLocaleTimeString('en-US', { hour12: true,
+                                             hour: "numeric",
+                                             minute: "numeric"}).toLowerCase()]
+}
+
+function absoluteTime(timeString){
+  let d = new Date(timeString);
+  return d.getTime();
+}
+
+function dateDivs(data, yoff){
+  var i;
+  let divs = [];
+  for (i=1; i<data.length; i++){
+    if(convertTime(data[i][0]["created_at"])[0]!==convertTime(data[i-1][0]["created_at"])[0]){
+      divs.push(<DateStyle offset={yoff[i]}>{convertTime(data[i][0]["created_at"])[0]}</DateStyle>);
+    }
+  }
+  console.log("date divs length is " + divs.length);
+  return divs;
+}
+
+function Timeline({courseCode, setTimeline}){
+  let [data, setData] = React.useState(null);
+  let [segLengths, setSegLengths] = React.useState(null);
+  let [yOffsets, setYOffsets] = React.useState(null);
+  let [loaded, setLoaded] = React.useState(false);
+  let [displayedCode, setDisplayedCode] = React.useState(null);
+  useEffect(()=>{
+      if(courseCode==null){
+        return;
+      }
+      setLoaded(false);
+      fetch(`https://penncourseplan.com/api/alert/statusupdate/${courseCode}`).then(res=>res.json()).then(result=>{
+        console.log(result)
+        result.sort((a,b)=>(a.created_at > b.created_at) ? 1 : -1);
+        let simplifiedData = result.reduce((ans, item, index) => { // preprocessing hte data
+          if(index==0){
+            return ans;
+          }
+          if(item["old_status"] == result[index-1]["old_status"]){
+            return ans;
+          }
+          if(item["old_status"] == "C" && item["new_status"] == "O"){
+            ans.push([item, "opened"]);
```

Comment:
```suggestion
            return ans.concat([[item, "opened"]]);
```
So we're not mixing immutable and mutable calls -- with a function like reduce it's better to stick to immutable operations that return a new list.